libraries/Microsoft.Bot.Builder.Dialogs/Memory/Scopes/SettingsMemoryScope.cs

```diff
@@ -73,91 +73,123 @@ public override void SetMemory(DialogContext dc, object memory)
-
-                            // if it's still a dictionary that's because we have a non-number value in it
-                            if (dict != null && dict.Count > 0)
-                            {
-                                // store using key as string
-                                node[lastKey] = child.Value;
-                            }
-                            else
-                            {
-                                // it's an array
-                                var arr = (object[])node;
-                                if (arr.Length < index + 1)
-                                {
-                                    // auto-resize to be bigger if we need to 
-                                    // NOTE: keys seem to be reverse indexed, which means we should have already "right-sized" the array
-                                    Array.Resize(ref arr, index + 1);
-                                    parent[parentKey] = arr;
-                                    node = arr;
-                                }
-
-                                // store using keyValue as an index
-                                arr[index] = child.Value;
-                            }
-                        }
-                        else
-                        {
-                            if (node is object[] arr)
-                            {
-                                // we got a non-number key but we have building an array, convert the array to dictionary
-                                // NOTE: keys is sorting reverse which means we should we always start out with a dictionary for nonnumberic keys
-                                var dict = new Dictionary<string, object>();
-                                for (var i = 0; i < arr.Length; i++)
-                                {
-                                    if (arr[i] != null)
-                                    {
-                                        dict[i.ToString(CultureInfo.InvariantCulture)] = arr[i];
-                                    }
-                                }
-
-                                parent[parentKey] = dict;
-                                node = dict;
-                            }
-
-                            node[lastKey] = child.Value;
-                        }
+                        currentNode = matchItem;
                     }
                 }
+
+                currentNode.Children.Add(new Node(child.Value));
             }
 
-            return settings;
+            return root;
+        }
+
+        private static object ConvertNodeToObject(Node node)
+        {
+            // If the child is leaf node, return its value directly.
+            if (node.Children.Count == 1 && node.Children[0].Children.Count == 0)
+            {
+                return node.Children[0].Value;
+            }
+
+            if (node.Children.All(u => int.TryParse(u.Value, out var number) && number >= 0))
```

Comment:
Can this be optimized to not loop through the children three times?

Reply:
I have merged the two "int.tryParse" and "int.parse" into one operator so that there only exists one time of the cast operator.
For the three loops of the array, personally think could be kept. The three loops are not nested, they are just the plain loops. If we forcibly turn it into a one-time operation, will need to change the original data or add additional memory space, and the most important is the readability will become very poor.

Just my opinion, no strong  point of view.